### ISSUE TITLE
Buffer overflow detected

### ISSUE THREAD

**Issue by mustakimur** [dianestephens/fileserver_cpp#35]
The client app has a buffer overflow which has been detected by the stack protector. Without the defense or with a compromised defense, a RCE is possible. The debug code shows the input sequence too:
```
(gdb) r
Starting program: /home/mustakim/projects/server/src/client 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
SERVER> Welcome to fileserver.

> connect guest
SERVERR> Welcome Guest User
> create user mmm
> User is already created. Please try again.
> login mmm
SERVERR> request encrypted data
SERVERR> request new encrypted data
SERVERR> Successful authentication
> ls
*** buffer overflow detected ***: terminated

Program received signal SIGABRT, Aborted.
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
50	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007ffff7bc9859 in __GI_abort () at abort.c:79
#2  0x00007ffff7c343ee in __libc_message (action=action@entry=do_abort, fmt=fmt@entry=0x7ffff7d5e07c "*** %s ***: terminated\n")
    at ../sysdeps/posix/libc_fatal.c:155
#3  0x00007ffff7cd6b4a in __GI___fortify_fail (msg=msg@entry=0x7ffff7d5e012 "buffer overflow detected") at fortify_fail.c:26
#4  0x00007ffff7cd53e6 in __GI___chk_fail () at chk_fail.c:28
#5  0x00007ffff7cd58c7 in __recv_chk (fd=<optimized out>, buf=<optimized out>, n=<optimized out>, buflen=<optimized out>, 
    flags=<optimized out>) at recv_chk.c:26
#6  0x000055555555a8eb in recv (__flags=0, __n=8, __buf=0x7fffffffcb50, __fd=3) at /usr/include/x86_64-linux-gnu/bits/socket2.h:42
#7  main () at client.cpp:419
```

In the source code, the vulnerable code is:
```c++
int bytesReceived = recv(sock, &filesize, sizeof(&filesize), 0);
if (bytesReceived == -1){
  cout << "There was an error getting response from server\r\n";
}
```

The `filesize` is a local int (4 bytes) and the `sizeof` is use `&filesize` which is an address (8 bytes), so a stack overflow is there.

**Comment by irfanOzn** [comment]
Dear Dr. Mustakimur,

Thank You for the bug report. To solve the stack overflow bug here, we replaced 64-bit sizeof(&filesize) length parameter  with sizeof(int) to make the length of the receive buffer equal to the bytes being read from the socket stream. Suggested change is this:


![image](https://user-images.githubusercontent.com/76975914/112742115-8881c780-8f59-11eb-892b-7be97024c473.png)



Thank You so much.
Sincerely Yours,
Team Red

**Comment by mustakimur** [comment]
Check the patch for the expected output of the command. If everything ok, commit the change and comment the commit number here.

**Comment by irfanOzn** [comment]
Dear Dr. Mustakimur,

Thank You so much. The bug has been solved in the commit number "9834c81aaef12c52440df96f3a4a977f4f4ccf88"


Sincerely Yours,
Team Red.

**Comment by dianestephens** [comment]
This issue was resolved and fix was committed. Thank you.
Resolved with commit "9834c81aaef12c52440df96f3a4a977f4f4ccf88"
